### ISSUE TITLE
Links within properties drawers of a node are not indexed

### ISSUE THREAD

**Issue by rdrg109** [org-roam/org-roam#1859]
### Table of Contents

1.  [Description](#orgce6e532)
    1.  [Prologue](#org5ff381b)
    2.  [Steps to reproduce (mwe)](#org6aae6f5)
    3.  [Backtrace](#orga8be334)
    4.  [Expected results](#org31fe73c)
    5.  [Actual results](#org68d2883)
    6.  [Related issues](#org68d2884)
2.  [Environment](#org95186ea)

<a id="orgce6e532"></a>

### Description

<a id="org5ff381b"></a>

#### Prologue

I usually reference nodes in the properties drawer of another node.   To understand why I do this, let me explain one of my uses of Org Mode: I use Org Mode for storing bookmarks. Each bookmark is stored as a headline (see below)

```org
* Org Roam
:PROPERTIES:
:LINK: https://www.orgroam.com/
:END:
```

I can then use the properties of the headline to store metadata of that bookmark

```org
* Org Roam
:PROPERTIES:
:LINK: https://www.orgroam.com/
:TOPIC+: emacs
:INSTANCE_OF: official site
:LANGUAGE: english
:END:
```

Two days ago, I had this idea of treating each entity, that is used as metadata as an Org Roam node, as a node so that the graph feature can also show relevant information on my bookmarks. So, a node would look like this

```org
* Org Roam
:PROPERTIES:
:LINK: https://www.orgroam.com/
:TOPIC+: [[id:96db2d4a-51c9-48fc-9f45-ac8d3331f314][emacs]]
:INSTANCE_OF: [[id:3b9fa6fa-67ea-4f96-b35d-4971e3a5999c][official site]]
:LANGUAGE: [[id:dc42cf9e-d7b6-4b5c-9463-abecececc89a][english]]
:ID: e395b878-a574-4c74-b89d-7c14522b2e1f
:END:
```

However, I noticed that the links located in properties drawers were not being indexed.

<a id="org6aae6f5"></a>

#### Steps to reproduce (mwe)

Instead of explaining the steps to reproduce, I'm going to present a minimal working example (mwe).

Let's suppose we have 4 nodes. The files containing these nodes are shown below.

```sh
echo '++++++++++++++++++++++++++++++ foo1'
cat ~/org-roam/20210917175708-foo1.org
echo '++++++++++++++++++++++++++++++ foo2'
cat ~/org-roam/20210917175721-foo2.org
echo '++++++++++++++++++++++++++++++ foo3'
cat ~/org-roam/20210917175724-foo3.org
echo '++++++++++++++++++++++++++++++ foo4'
cat ~/org-roam/20210917182503-foo4.org
```

```
++++++++++++++++++++++++++++++ foo1
:PROPERTIES:
:ID: 27c9c572-f2da-45f2-901c-1a44f76536d3
:END:
#+title: foo1

A simple paragraph. The following is a link [[id:eacfc739-5909-4432-9ef5-0b758811a0e7][foo3]].
++++++++++++++++++++++++++++++ foo2
:PROPERTIES:
:ID: 7a1fe672-899d-4f83-9e72-2875ac547ee5
:END:
#+title: foo2

A simple paragraph. The following is a link [[id:eacfc739-5909-4432-9ef5-0b758811a0e7][foo3]].
++++++++++++++++++++++++++++++ foo3
:PROPERTIES:
:ID: eacfc739-5909-4432-9ef5-0b758811a0e7
:RANDOM_NAME_1: [[id:27c9c572-f2da-45f2-901c-1a44f76536d3][foo1]]
:RANDOM_NAME_2: [[id:7a1fe672-899d-4f83-9e72-2875ac547ee5][foo2]]
:END:
#+title: foo3

[[id:6ab19f71-2315-47fd-9497-3f7f48cde5af][foo4]]
++++++++++++++++++++++++++++++ foo4
:PROPERTIES:
:ID: 6ab19f71-2315-47fd-9497-3f7f48cde5af
:END:
#+title: foo4

```

We can see that the nodes are correctly indexed by executing the following command.

```lisp
(pp (org-roam-db-query [:select * :from nodes]))
```

```
(("27c9c572-f2da-45f2-901c-1a44f76536d3" "/home/beep1560/org-roam/20210917175708-foo1.org" 0 1 nil nil nil nil "foo1"
  (("CATEGORY" . "20210917175708-foo1")
   ("ID" . "27c9c572-f2da-45f2-901c-1a44f76536d3")
   ("BLOCKED" . "")
   ("FILE" . "/home/beep1560/org-roam/20210917175708-foo1.org")
   ("PRIORITY" . "B"))
  nil)
 ("7a1fe672-899d-4f83-9e72-2875ac547ee5" "/home/beep1560/org-roam/20210917175721-foo2.org" 0 1 nil nil nil nil "foo2"
  (("CATEGORY" . "20210917175721-foo2")
   ("ID" . "7a1fe672-899d-4f83-9e72-2875ac547ee5")
   ("BLOCKED" . "")
   ("FILE" . "/home/beep1560/org-roam/20210917175721-foo2.org")
   ("PRIORITY" . "B"))
  nil)
 ("6ab19f71-2315-47fd-9497-3f7f48cde5af" "/home/beep1560/org-roam/20210917182503-foo4.org" 0 1 nil nil nil nil "foo4"
  (("CATEGORY" . "20210917182503-foo4")
   ("ID" . "6ab19f71-2315-47fd-9497-3f7f48cde5af")
   ("BLOCKED" . "")
   ("FILE" . "/home/beep1560/org-roam/20210917182503-foo4.org")
   ("PRIORITY" . "B"))
  nil)
 ("eacfc739-5909-4432-9ef5-0b758811a0e7" "/home/beep1560/org-roam/20210917175724-foo3.org" 0 1 nil nil nil nil "foo3"
  (("CATEGORY" . "20210917175724-foo3")
   ("RANDOM_NAME_2" . "[[id:7a1fe672-899d-4f83-9e72-2875ac547ee5][foo2]]")
   ("RANDOM_NAME_1" . "[[id:27c9c572-f2da-45f2-901c-1a44f76536d3][foo1]]")
   ("ID" . "eacfc739-5909-4432-9ef5-0b758811a0e7")
   ("BLOCKED" . "")
   ("FILE" . "/home/beep1560/org-roam/20210917175724-foo3.org")
   ("PRIORITY" . "B"))
  nil))
```

Now, let's see what is stored in the `links` table

```sqlite
SELECT * FROM links;
```

```
pos  source                                  dest                                    type  properties    
---  --------------------------------------  --------------------------------------  ----  --------------
121  "27c9c572-f2da-45f2-901c-1a44f76536d3"  "eacfc739-5909-4432-9ef5-0b758811a0e7"  "id"  (:outline nil)
121  "7a1fe672-899d-4f83-9e72-2875ac547ee5"  "eacfc739-5909-4432-9ef5-0b758811a0e7"  "id"  (:outline nil)
209  "eacfc739-5909-4432-9ef5-0b758811a0e7"  "6ab19f71-2315-47fd-9497-3f7f48cde5af"  "id"  (:outline nil)
```

As we can see above, only the following links are included in the `links` table

-   `foo1` to `foo3`
-   `foo2` to `foo3`
-   `foo3` to `foo4`

However, the following links are not.

-   `foo3` to `foo1`
-   `foo3` to `foo2`

Note that the links that weren't indexed were the ones located in the properties drawer of the `foo3` node.

<a id="orga8be334"></a>

#### Backtrace

This bug report doesn't have a backtrace. A minimal working example has been provided in "Steps to reproduce"

<a id="org31fe73c"></a>

#### Expected results

Links within properties drawers of a node are indexed.

<a id="org68d2883"></a>

#### Actual results

Links within properties drawers of a node aren't indexed.

<a id="org68d2884"></a>

#### Related issues

An issue on this had been created before: #1464 

<a id="org95186ea"></a>

### Environment

```text
- Emacs: GNU Emacs 27.2 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.27, cairo version 1.17.4)
 of 2021-03-26
- Framework: N/A
- Org: Org mode version 9.4.4 (release_9.4.4 @ /usr/share/emacs/27.2/lisp/org/)
- Org-roam: 2.1.0
```

**Comment by rdrg109** [comment]
I've been doing some research and found that the root cause of this problem is `org-element-map` (which is used by some functions) because it iterates through all the links except for the ones within properties drawers.


# Root cause of the problem

The function `org-roam-db-autosync--try-update-on-save-h` is included in `after-save-hook` and calls `org-roam-db-update-file` (see below).

<https://github.com/org-roam/org-roam/blob/1795039ab93ef19611dbb3fca21c4211c4e655a9/org-roam-db.el#L628-L630>

<https://github.com/org-roam/org-roam/blob/1795039ab93ef19611dbb3fca21c4211c4e655a9/org-roam-db.el#L632-L634>

The function `org-roam-db-update-file` calls `org-roam-db-map-links` (see below).

<https://github.com/org-roam/org-roam/blob/1795039ab93ef19611dbb3fca21c4211c4e655a9/org-roam-db.el#L512-L514>

The function `org-roam-db-map-links` calls `org-element-map` (see below). This is done to iterate through all the links in the current buffer.

<https://github.com/org-roam/org-roam/blob/1795039ab93ef19611dbb3fca21c4211c4e655a9/org-roam-db.el#L294-L301>

And `org-element-map` doesn't consider links within properties drawers. Proof on this is shown in the next section.

# Proof that `org-element-map` doesn't consider links within properties drawers

The following example shows that all links in the buffer are considered by `org-element-map`.

```lisp
(with-temp-buffer
  (org-mode)
  (let ((lines '("[[1]]"
                 "[[2]]"
                 "[[3]]")))
    (insert (string-join lines "\n")))
  (org-element-map (org-element-parse-buffer)
      'link
    (lambda (link) (org-element-property :raw-link link))))
```

```
("1" "2" "3")
```

The following example shows that all links except for the ones located within the properties drawer are not considered by `org-element-map`.


```lisp
(with-temp-buffer
  (org-mode)
  (let ((lines '("* John Doe"
                 ":PROPERTIES:"
                 ":FOO: [[1]]"
                 ":BAR: [[2]]"
                 ":FIZZ: [[3]]"
                 ":END:"
                 "[[4]]"
                 "[[5]]")))
    (insert (string-join lines "\n")))
  (org-element-map (org-element-parse-buffer)
      'link
    (lambda (link) (org-element-property :raw-link link))))
```

```
("4" "5")
```

**Comment by rdrg109** [comment]
Another workaround would be to make `org-roam-db-map-links` include the links in a property drawer. This would imply not using `org-element-map` because as was mentioned before, the current implementation of `org-element-map` doesn't consider links in property drawers.

**Comment by rdrg109** [comment]
A issue on this topic had been created before #1464

**Comment by jethrokuan** [comment]
Closed with #1868
